### DISCUSSION TITLE
Pylance not autocompleting metaclass/init_subclass keyword arguments

### DISCUSSION THREAD

**Discussion by StockerMC** [microsoft/pylance-release#2699] [Enhancement]
## Environment data
-   Language Server version: v2021.10.1
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.9.7

## Expected behaviour
For the keyword arguments defined in `__init_subclass__` and `__new__` to be autocompleted. Due to https://github.com/microsoft/pyright/issues/2436, the metaclass keyword arguments not being suggested may not be a bug.
Instead of the keyword arguments of `__init_subclass__`/`__new__` being suggested, the attributes of the class are suggested.

## Actual behaviour
![image](https://user-images.githubusercontent.com/44980366/137502339-41525e46-834b-4d62-9e10-7e51dd994b8c.png)
![image](https://user-images.githubusercontent.com/44980366/137502358-96dd3546-c211-4cd5-bcb1-a2f098976e43.png)

## Code Snippet / Additional information
```python
from typing import TypeVar, Tuple, Dict, Any, Type, TYPE_CHECKING

CMT = TypeVar('CMT', bound='CommandMeta')

class BaseCommand:
    if TYPE_CHECKING:
        __foo__: int
        __bar__: str

    def __init_subclass__(cls, *, foo: int, bar: str) -> None: ...

class CommandMeta(type):
    if TYPE_CHECKING:
        __foo__: int
        __bar__: str

    def __new__(
        cls: Type[CMT],
        cls_name: str,
        bases: Tuple[type, ...],
        attrs: Dict[str, Any],
        *,
        foo: int,
        bar: str,
    ) -> CMT: ...
```

**Comment by erictraut** [comment]
This is a very niche use case for auto-complete — especially the `__new__` case, which pyright (and mypy) do not support. Even `__init_subclass__` and custom keyword arguments for class declarations are rarely used in typical Python programs.

**Comment by StockerMC** [comment]
Is there a reason as to why it suggests attributes from the class though (as shown in the screenshots)?

**Comment by erictraut** [comment]
That's a good point. It's currently using the class's scope to build the list of suggestions, but the arguments for the class are evaluated within the containing scope (in this case, the module scope). I've fixed that issue so the correct scope is used in this context.

**Comment by StockerMC** [comment]
Thanks.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
